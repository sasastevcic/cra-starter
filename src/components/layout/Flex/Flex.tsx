/* eslint-disable @typescript-eslint/ban-ts-comment */
import { forwardRef } from 'react';
import { FlexProps } from './Flex.data';
import { StyledFlex } from './Flex.styles';

export const Flex = forwardRef<Element, FlexProps>(
	(
		{
			flexDirection,
			justifyContent,
			alignItems,
			flexWrap,
			flexGrow,
			flexShrink,
			children,
			elementType = 'div',
			...props
		},
		ref,
	) => (
		<StyledFlex
			data-testid="Flex"
			ref={ref}
			// @ts-ignore
			as={elementType}
			$flexDirection={flexDirection}
			$justifyContent={justifyContent}
			$alignItems={alignItems}
			$flexWrap={flexWrap}
			$flexGrow={flexGrow}
			$flexShrink={flexShrink}
			{...props}
		>
			{children}
		</StyledFlex>
	),
);
