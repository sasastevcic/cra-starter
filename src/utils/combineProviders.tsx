/* eslint-disable @typescript-eslint/no-explicit-any */
import { ComponentProps, ComponentType, FunctionComponent, ReactNode } from 'react';

type Providers = Array<[ComponentType<any>, ComponentProps<any>?]>;

export const combineProviders = (
	providers: Providers,
): FunctionComponent<{ children: ReactNode }> =>
	providers.reduce(
		(AccumulatedProviders, [Provider, props = {}]) =>
			({ children }) =>
				(
					<AccumulatedProviders>
						<Provider {...props}>{children}</Provider>
					</AccumulatedProviders>
				),
		({ children }) => <>{children}</>,
	);
