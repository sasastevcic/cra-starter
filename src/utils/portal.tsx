import { ReactNode, ReactPortal, useMemo } from 'react';
import { createPortal } from 'react-dom';

interface Portal {
	portalId: string;
	children: ReactNode;
}

const Portal = ({ portalId, children }: Portal): ReactPortal => {
	const portalElement = useMemo(() => document.getElementById(portalId), [portalId]);

	if (!portalElement) {
		throw new Error(`Element not found! Element id: ${portalId}`);
	}

	return createPortal(children, portalElement);
};

export default Portal;
