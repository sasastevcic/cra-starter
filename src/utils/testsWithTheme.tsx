/* eslint-disable @typescript-eslint/ban-ts-comment */
import { ReactElement, ComponentType, ReactNode } from 'react';
import { render, RenderOptions, RenderResult } from '@testing-library/react';
import { ThemeProvider } from 'styled-components';
import { theme } from '../styles/config/theme';

export const componentWithTheme = (component: ReactNode): ReactElement => (
	// @ts-ignore
	<ThemeProvider theme={theme}>{component}</ThemeProvider>
);

export const renderWithTheme = (component: ReactElement, options?: RenderOptions): RenderResult => {
	const wrapper: ComponentType<{ children: ReactNode }> = ({ children }) =>
		componentWithTheme(children);

	return render(component, { wrapper, ...options });
};
