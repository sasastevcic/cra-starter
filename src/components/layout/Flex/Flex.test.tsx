import renderer from 'react-test-renderer';
import { renderWithTheme } from '../../../utils/testsWithTheme';
import Noop from '../../atoms/Noop';
import Flex from './index';

describe('Flex', () => {
	describe('Render', () => {
		it('should render', () => {
			const { getByTestId } = renderWithTheme(
				<Flex>
					<Noop />
				</Flex>,
			);

			const element = getByTestId('Flex');

			expect(element).toBeInTheDocument();
		});

		it('should render header tag', () => {
			const { getByTestId } = renderWithTheme(
				<Flex elementType="header">
					<Noop />
				</Flex>,
			);

			const element = getByTestId('Flex');

			expect(element.tagName).toBe('HEADER');
		});

		it('should render correctly', () => {
			const tree = renderer
				.create(
					<Flex>
						<Noop />
					</Flex>,
				)
				.toJSON();

			expect(tree).toMatchSnapshot();
		});
	});

	describe('Content', () => {
		it('should have elements as children', () => {
			const { getByTestId } = renderWithTheme(
				<Flex>
					<span>Child</span>
				</Flex>,
			);

			const element = getByTestId('Flex');
			const child = element.querySelector('span');

			expect(element).toContainElement(child);
		});
	});
});
