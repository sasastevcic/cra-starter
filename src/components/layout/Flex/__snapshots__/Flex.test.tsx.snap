// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Flex Render should render correctly 1`] = `
<div
  className="sc-bczRLJ bsZsBM"
  data-testid="Flex"
/>
`;
